### ISSUE TITLE
[CT-3386] [Feature] Option for full refresh just a single model

### ISSUE THREAD

**Issue by aBBDnGus** [dbt-labs/dbt-core#9082]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Currently, there is an option to fully refresh all incremental models via `dbt run --full-refresh`. This executes the full refresh on all selected incremental models. But there is no option to build all models, but to fully refresh only one of them.

Minimal Example:
I have the following models:
* `model_1`: view
* `model_2`: incremental
* `model_3`: incremental, depends on `model_1` and `model_2`
* `model_4`: view, depends on `model_3`
* `model_5`: incremental, depends on `model_3`

Now, the models `model_1` and `model_3` are changed.
* If I run `dbt run --full-refresh`, then all incremental models (i.e. `model_2`, `model_3` and `model_5`) are fully refreshed. This executes the refresh on `model_2` and `model_5` unnecessarily.
* If I run `dbt run --full-refresh --models model_3`, then `model_3` is fully refreshed, but the changes of `model_1` are not applied and `model_4` is deleted.

### Describe alternatives you've considered

Option 1: Perform the following commands:
```
dbt run --models +model_3
dbt run --models model_3 --full-refresh
dbt run
```

Option 2: Temporarily change `dbt_project.yml` and perform `dbt run`
```
models:
  +full_refresh: false
  model_3:
    full_refresh: true
```

### Who will this benefit?

Projects with multiple incremental models, if the incremental models change from time to time and it is expensive to fully refresh incremental models.

**Comment by jaypeedevlin** [comment]
This particular example could be solved by

```
dbt run -s +model_3 --full-refresh
```

**Comment by jaypeedevlin** [comment]
Additionally the `model_4` being deleted seems like binding views in redshift which can be disabled via config.

**Comment by dbeatty10** [comment]
Thanks for reaching out about this @aBBDnGus !

Can you let us know if this suggestion from @jaypeedevlin works for you or not?

```shell
dbt run -s +model_3 --full-refresh
```

**Comment by aBBDnGus** [comment]
Thanks for your answers!
* `dbt run -s +model_3 --full-refresh` does not work, since then also the incremental table `model_2` is fully refreshed. I want to avoid this.
* I'm using PostgreSQL and unfortunately there are not late binding views.

**Comment by dbeatty10** [comment]
Here's a minimal project that replicates the scenario you are describing:

<details>

<summary>Project model files</summary>

`models/model_1`

```sql
{{ config(materialized="view") }}

select 1 as id
```

`models/model_2`

```sql
{{ config(materialized="incremental") }}

select 1 as id
```

`models/model_3`

```sql
{{ config(materialized="incremental") }}

-- depends on: {{ ref("model_1") }}
-- depends on: {{ ref("model_2") }}
select 1 as id
```

`models/model_4`

```sql
{{ config(materialized="view") }}

-- depends on: {{ ref("model_3") }}
select 1 as id
```

`models/model_5`

```sql
{{ config(materialized="incremental") }}

-- depends on: {{ ref("model_3") }}
select 1 as id
```

`models/model_6`

```sql
{{ config(materialized="incremental") }}

select 1 as id
```

</details>

In order to demonstrate the case where there are subgraphs that don't involve `model_3`, I added one extra incremental model above named `model_6`.

Here are a few ways to full-refresh just a single model, while also running the rest of the models using their default configurations for `full_refresh`:

### Option 3

Add `full_refresh` to the model-level `config` macro:

`models/model_3.sql`

```yaml
{{ config(
        materialized="incremental",
        full_refresh=True,
    )
}}

```

Then simply run (or build) everything:

```shell
dbt run
```

### Option 2

Like you mentioned in your alternatives, add a model-level config within `dbt_project.yml`:

`dbt_project.yml`

```yaml
name: "my_project"

# ...

models:
  my_project:
    model_3:
      full_refresh: true
```

Then run everything:

```shell
dbt run
```

### Option 1

Run all models (including model_3), then full refresh model_3, then run all models:
```shell
dbt run --models +model_3
dbt run --models model_3 --full-refresh
dbt run
```

The trade-off is that model_3 is executed three times and anything upstream of it is executed two times. Everything else is executed a single time.

### Option 0

Split the DAG into three parts that adhere to MECE and run one command per partition so that everything only runs a single time:

```shell
dbt run --models +model_3 --exclude model_3
dbt run --models model_3 --full-refresh
dbt run --exclude +model_3
```

There are a few variants on this idea wrap these commands so they can be executed with less keystrokes. See below for a couple.

#### Parameterized function alias in zsh

Without loss of generalization, we can create a [parameterized function alias](https://linuxhandbook.com/zsh-aliases/#create-function-aliases-on-zsh) within `.zshrc` that contains the desired commands:

```
function full_refresh_model() {
  # Parameterize the model name
  local model_name="$1"

  # Define the commands with the parameterized model name
  dbt run --models +$model_name --exclude $model_name
  dbt run --models $model_name --full-refresh
  dbt run --exclude +$model_name
}
```

After restarting the shell (or sourcing the `.zshrc` file), then we can run it like this:

```shell
full_refresh_model model_3
```

#### Programmatic invocations

```python
import sys

from dbt.cli.main import dbtRunner, dbtRunnerResult


def full_refresh_model(model_name):
    # initialize
    dbt = dbtRunner()

    invocations = []

    upstream = f"+{model_name}"

    invocations.append(["run", "--models", upstream, "--exclude", model_name])
    invocations.append(["run", "--models", model_name, "--full-refresh"])
    invocations.append(["run", "--exclude", upstream])

    # run the commands for each invocation
    for cli_args in invocations:
        dbt.invoke(cli_args)


if __name__ == "__main__":
    if len(sys.argv) > 1:
        model_name = sys.argv[1]
        full_refresh_model(model_name)
    else:
        print("No model name provided")

```

```shell
python runner.py model_3
```

### Summary

Of the options listed above, I'd personally go for **Option 3** because it's simple to add the config to the applicable model and trivial to execute the desired nodes. And then it's easy to undo when finished.

Could we create some kind of special syntax that allows a command-line arg that designates just a single model out of many to be full-refreshed? We surely could. It might be something like this:

```
dbt run --select {any selection you want here} --only-full-refresh model_3
```

The trade-off would be adding more complication during node selection and another thing write documentation for.

But do we _need_ to?

This doesn't feel like a necessity to me, and we have a collection of alternatives listed above.

They are all at least possible and **Option 3** is almost trivial (once you know about it).

So I'm going to close this as "not planned". If we get feedback that this is a highly impactful problem from many people, we can consider re-opening this issue.

In the meantime, we could document Option 3 as an example under the heading "Full refresh just a single model" (or something similar).

**Comment by aBBDnGus** [comment]
Thanks for the answers and clarifying! It would totally appreciate if you add this example to your documentation.
I personally prefer **Option 0**. The dbt models are integrated in our workflow and deployment. It wouldn't make any sense, if we modify a model, then deploy it (with `dbt run`) and then modify it again (with removing `full-refresh=True`).

**Comment by dbeatty10** [comment]
> It would totally appreciate if you add this example to your documentation.

Created this issue to add example(s) to our documentation: https://github.com/dbt-labs/docs.getdbt.com/issues/4504

### Follow-up question

To help understand the details of why you prefer Option 0, what type of interface would you use to add the Option 0 commands? Are these stored in source control for you somewhere? Or are they in a web interface or CLI?

Option 0 commands:
```shell
dbt run --models +model_3 --exclude model_3
dbt run --models model_3 --full-refresh
dbt run --exclude +model_3
```

**Comment by jtcohen6** [comment]
@sdurry had an idea that might solve for this (and also https://github.com/dbt-labs/dbt-core/issues/4473): Create a [new model version](https://docs.getdbt.com/docs/collaborate/govern/model-versions).

Your new model version has a new physical location in the DWH, so it rebuilds from scratch. Meanwhile, the downstream references still resolve to the same upstream model. You can also split this into two PRs: the first to implement the new model version, the second to migrate over to it (by marking as "latest" and/or removing the previous model version).

**Comment by aBBDnGus** [comment]
@dbeatty10 Currently, our deployment process is as follows: We merge a pull request and then the CI tool automatically executed `dbt run`. If some special migration is necessary (e.g. full-refresh of a model), we execute this migration manually. In future, we plan to store also the migrations in our source control.

**Comment by FredericoCoelhoNunes** [comment]
This functionality would be really helpful for us; we are running dbt from Airflow, and have different full-refresh schedules for different models (we want to full-refresh very large models monthly, and other weekly, for example).

Instead of having to create multiple Airflow DAGs to account for the different full-refresh schedules, we would love to be able to pass which models to full-refresh from Airflow to dbt. This way Airflow handles the orchestration, and dbt makes sure that the entire graph is run in the right order (something that became more difficult to keep track of when we started splitting our dbt run into several different runs).
